client.go

```diff
@@ -143,74 +143,86 @@ func (c *client) decodeMeta(data []byte) (*proto.ServerName, error) {
 func (c *client) init() error {
 	zkclient, _, err := zk.Connect(c.zkHosts, time.Second*30)
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
 	c.zkClient = zkclient
 
 	res, _, _, err := c.zkClient.GetW(c.zkRoot + zkRootRegionPath)
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
+
 	c.rootServerName, err = c.decodeMeta(res)
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
+
 	log.Debug("connect root region server...", c.rootServerName)
 	conn, err := newConnection(serverNameToAddr(c.rootServerName), false)
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
+
 	// set buffered regionserver conn
 	c.cachedConns[serverNameToAddr(c.rootServerName)] = conn
 
 	res, _, _, err = c.zkClient.GetW(c.zkRoot + zkMasterAddrPath)
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
+
 	c.masterServerName, err = c.decodeMeta(res)
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
+
 	return nil
 }
 
 // get connection
-func (c *client) getConn(addr string, isMaster bool) *connection {
+func (c *client) getConn(addr string, isMaster bool) (*connection, error) {
 	c.mu.RLock()
-	if s, ok := c.cachedConns[addr]; ok {
-		defer c.mu.RUnlock()
-		return s
-	}
+	conn, ok := c.cachedConns[addr]
 	c.mu.RUnlock()
 
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	conn, err := newConnection(addr, isMaster)
+	if ok {
+		return conn, nil
+	}
+
+	var err error
+	conn, err = newConnection(addr, isMaster)
```

Comment:
If this function is concurrently accessed, we may create several `conn`s for a single `addr`. Only one `conn` can be added to cache, others are returned only once and will never be closed.